lib/services/hpcpackage_remain.pm

```diff
@@ -0,0 +1,86 @@
+# SUSE's openQA tests
+#
+# Copyright © 2020 SUSE LLC
+#
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.  This file is offered as-is,
+# without any warranty.
+
+# Summary: Make installed versions of HPC library packages remain 
+# installed after migration to new service pack
+# Steps: 1.Install previous SP (SLE-HPC text mode profile)
+# 2. Make a list of installed packages ('rpm -qa')
+# 3. Install all 'library wrappers' (packages matching -hpc, not 
+#    having a version number with '_' after their name).
+# 4. Make a list of installed packages and diff against list from 2.
+# 5. From the diff extract the list of HPC library packages (packages
+#    matching *-hpc*) with 'underscored' version numbers after the 
+#    package name) and save this list.
+# 6. Perform a migration
+# 7. Confirm all packages in the list generated in 5. remain installed.
+# Maintainer: Yutao Wang <yuwang@suse.com>
+
+package services::hpcpackage_remain;
+use base 'opensusebasetest';
+use testapi;
+use utils;
+use strict;
+use Data::Dumper;
+use warnings;
+
+my @diffpkg;
+
+sub list_pkg {
```

Comment:
I'd suggest to use 2 array to hold the pre_list and instlled_list then compare these 2 arrays like get_pattern_list and get_pattern_list in lib/util.pm. just my 2 cents.

Reply:
At here, I think same logic can use same function. Don't need make a new function.